### PR TITLE
Account for 'appProp' in 'reduceNonRepPrim'. Fix #974

### LINKED ISSUE #974: `reduceNonRepPrim` potentially introduces illegal variables
Consider the primitive reduction of `map`:

https://github.com/clash-lang/clash-compiler/blob/33047ee3a81e5c1551df7e8c9c05679efcf0e9b6/clash-lib/src/Clash/Normalize/PrimitiveReductions.hs#L141-L171

`extractElems` is called with `is0`: the `InScopeSet` of the current transformation. Given the pseudocode:

```haskell
map (\x -> f 3 x) xs :: Vec 2 a
```

the primitive reduction might generate:

```haskell
let 
  x = xs !! 0
  y = xs !! 1
in 
     (\x -> f 3 x) x 
  :> (\x -> f 3 x) y 
  :> Nil
```

This is incorrect as `reduceNonRepPrim` is potentially followed by an `appProp`:

https://github.com/clash-lang/clash-compiler/blob/fc8beb3edba50c14de3084cfa571f83235ee2423/clash-lib/src/Clash/Normalize.hs#L355-L363

`appProp` has an invariant assuming:

https://github.com/clash-lang/clash-compiler/blob/02616d7a97612d123b42172f628f1f8f80e86142/clash-lib/src/Clash/Normalize/Transformations.hs#L1095-L1097

`reduceNonRepPrim` should therefore make sure it doesn't introduce any new variables already mentioned in the body of `fun1` (as well as the "normal" in scope set).

### PR DESCRIPTION
**Pull request by martijnbastiaan**

### CODE REVIEW

File: clash-lib/src/Clash/Core/Term.hs [hunk 1]
```diff
@@ -241,3 +243,47 @@ partitionTicks
...
+  TyApp t1 _ -> pure (f t1)
+  Letrec bndrs t1 -> pure (f t1) <> mconcat (map (walkTerm' f . snd) bndrs)
+  Case t1 _ alts -> pure (f t1) <> mconcat (map (walkTerm' f . snd) alts)
+  Cast t1 _ _ -> walkTerm' f t1
+  Tick _ t1 -> pure (f t1)
+
+-- | Visit all terms in a term, testing it with a predicate, and returning
+-- a list of predicate yields.
+walkTerm :: (Term -> Maybe a) -> Term -> [a]
+walkTerm f t = catMaybes (DList.toList (walkTerm' f t))
```
[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply clash-lib/src/Clash/Core/Term.hs]: 1. Agreed.

2. Hmm, I'm not a fan of point-free notation _generally_. It makes it harder to use debug/trace functions (as it's not let-bound, so you can't easily do preprocessing before printing), harder to parse for beginners, and it obscures argument names. I'd rather have us only use it for HO functions and "bootstrapping" situations, but not for top-level definitions. As you said, it's a point of contention for Haskellers, so I'm not going to put my foot down. That being said, I'm fine with changing it for this function as it's small enough and can't really have descriptive argument names anyway.

[the remark this answers is not in the document]
Comment (alex-mckenna) [reply clash-lib/src/Clash/Core/Term.hs]: I hadn't considered the effect on debugging / tracing before, good point. Again, I think it's a thing to go back and codify better in the style guide (which currently just says not to overuse it).

Maybe 2020 will be the year of beautifully formatted `clash-compiler` :wink:

File: clash-lib/src/Clash/Core/Term.hs [hunk 2]
```diff
@@ -241,3 +243,47 @@ partitionTicks
...
+
+-- Collect all term ids mentioned in a term
+collectTermIds :: Term -> [Id]
+collectTermIds t = concat (walkTerm (Just . go) t)
```
[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply clash-lib/src/Clash/Core/Term.hs]: Alright, let's discuss this later. To foreshadow :P, my experience with automated checkers hasn't been that good. I _loath_ going through a CI cycle just to find out I missed a space / ETA reduction somewhere.

**Comment by martijnbastiaan** [comment]
It's really tricky to write a test for this as it depends on `extractElems` generating the "right" (wrong?) uniques in the first place. That said, `appProp` should check whether its invariants are violated when Clash is compiled with debugging options enabled because there's bound to be more transformations that mess this up. I'll add an issue.

**Comment by alex-mckenna** [comment]
All right, if you make an issue for that and reference it back to this then I'll approve

**Comment by martijnbastiaan** [comment]
@alex-mckenna I've made an issue for `appProp` invariant checking and applied your suggestions.

**Pull request merged**